nevergrad/optimization/optimizerlib.py

```diff
@@ -1425,32 +1425,126 @@ def _internal_tell_not_asked(self, candidate: p.Parameter, value: float) -> None
+        parent = self.parents[self.num_ask % len(self.parents)]
+        candidate = parent.spawn_child().set_standardized_data(individual, reference=self.parametrization)
+        if parent is self.parametrization:
+            candidate.heritage["lineage"] = candidate.uid
+        candidate._meta["sigma"] = self.sigma
+        return candidate
 
     def _internal_tell_candidate(self, candidate: p.Parameter, value: float) -> None:
         candidate._meta["loss"] = value
-        self.popsize.add_value(value)
         self.children.append(candidate)
-        if len(self.children) >= self.popsize.llambda:
+        if len(self.children) >= self.llambda:
             # Sorting the population.
             self.children.sort(key=lambda c: c._meta["loss"])
             # Computing the new parent.
-            mu = self.popsize.mu
-            self.current_center = sum(c.get_standardized_data(reference=self.parametrization)  # type: ignore
-                                      for c in self.children[: mu]) / mu
-            t1 = [(c.get_standardized_data(reference=self.parametrization) - self.current_center)**2 for c in self.children[: mu]]
+            self.parents = self.children[: self.mu]
+            self.children = []
+            self.current_center = sum(c.get_standardized_data(reference=self.parametrization) for c in self.parents) / self.mu  # type: ignore
             # EMNA update
-            self.sigma = np.sqrt(sum(t1) / (mu))
-            imp = max(1, (np.log(self.popsize.llambda) / 2)**(1 / self.dimension))
-            if self.num_workers / self.dimension > 16:
+            stdd = [(self.parents[i].get_standardized_data(reference=self.parametrization) - self.current_center)**2 for i in range(self.mu)]
+            self.sigma = np.sqrt(sum(stdd) / (self.mu * self.dimension))
+            if self.num_workers / self.dimension > 16: # faster decrease of sigma if large parallel context
+                imp = max(1, (np.log(self.llambda) / 2)**(1 / self.dimension))
                 self.sigma /= imp
-            self.parents = self.children[: mu]
+
+    def _internal_tell_not_asked(self, candidate: p.Parameter, value: float) -> None:
+        base.TellNotAskedNotSupportedError
+
+
+@registry.register
+class NaiveEMNA(EMNA):
+    """Simple Estimation of Multivariate Normal Algorithm (EMNA).
+       Noiseless implementation
+    """
+
+    def _internal_provide_recommendation(self) -> ArrayLike:
+        return self.current_bests["optimistic"].x  # Naive version
+
+
+@registry.register
+class AnisoEMNA(EMNA):
+    """Simple Anisotropic Estimation of Multivariate Normal Algorithm (EMNA).
+    """
+
+    # pylint: disable=too-many-instance-attributes
+
+    def __init__(self, parametrization: IntOrParameter, budget: Optional[int] = None, num_workers: int = 1) -> None:
+        super().__init__(parametrization, budget=budget, num_workers=num_workers)
+        self.sigma = np.ones(self.dimension)
```

Comment:
the typing issue is here because you're setting a np.ndarray to `sigma` which is a float in the parent class.
You'd need to type it to `sigma: tp.Union[float, np.ndarray]` in the **parent** class

Reply:
Ok thx for all comments ; I try to correct it quickly, some stuff are easy to change ; the correction of code duplication is a good idea, I will look into that (it is not intuitive right now, as I haven't looked deeply enough in the parametrization but I will work on it :).

Reply:
Let me know if you need help, at some point I can make a PR on top of your PR to suggest changes if you need

Reply:
That's great!
There's a second step to make it work:
You'll need to add a `ConfiguredOptimizer` and register variants like it's done here for OnePlusOne:
https://github.com/facebookresearch/nevergrad/blob/e07ecefe0c3a7fdffe72abdef4eb0abf7ea7da11/nevergrad/optimization/optimizerlib.py#L117-L166

For you that would mean this:
```python
class ConfiguredEMNA(base.ConfiguredOptimizer):
    """TODO
    """

    # pylint: disable=unused-argument
    def __init__(
        self,
        *,
            isotropic: bool = True,
            naive: bool = True
    ) -> None:
        super().__init__(EMNA, locals())
```

Actually if you are OK, I would call this new class `EMNA`, and the one you did `_EMNA` to hide it since the API that we need to document is the one in the new class.
The syntax is then for instance `EMNA(isotropic=False)(parametrization, budget)` to instantiate an optimizer.
You can preregister variants, like:
```python
NaiveIsoEMNA = EMNA().set_name("NaiveIsoEMNA", register=True)
NaiveAnisoEMNA = EMNA(isotropic=False).set_name("NaiveAnisoEMNA", register=True)
IsoEMNA = EMNA(naive=False).set_name("IsoEMNA", register=True)
AnisoEMNA = EMNA(inaive=False, sotropic=False).set_name("AnisoEMNA", register=True)
```

Then a few things to consider:
- please add a docstring in the new class, it will be automatically visible in the documentation when it's merged. What could be of interest for users is when this algorithm should be used (which cases will it be good at? bad at?) and what is the impact of each parameter
- add the new configured optimizer to optimization/families.py
- I'm trying to limit the number of variants in optimizerlib (work in progress...), to keep only the most useful variants for user, if you can leave only one or two there it would be great, the others should be added to optimization/experimentalvariants.py so that they can be straightforwardly used in the benchmarks.

Reply:
Hello, thx a lot for your remarks. I think I have taken into account everything, but do not hesitate to tell me if something is not correct.